pkg/storage/ingest/reader.go

```diff
@@ -0,0 +1,331 @@
+	var (
+		now        = time.Now()
+		numRecords = 0
+	)
+
+	fetches.EachRecord(func(record *kgo.Record) {
+		numRecords++
+		r.metrics.receiveDelay.Observe(now.Sub(record.Timestamp).Seconds())
+	})
+
+	r.metrics.recordsPerFetch.Observe(float64(numRecords))
+}
+
+func (r *PartitionReader) newKafkaReader(at kgo.Offset) (*kgo.Client, error) {
+	client, err := kgo.NewClient(
+		kgo.ClientID(r.kafkaClientID),
+		kgo.SeedBrokers(r.kafkaAddress),
+		kgo.ConsumePartitions(map[string]map[int32]kgo.Offset{
+			r.kafkaTopic: {r.partitionID: at},
+		}),
+		kgo.FetchMinBytes(1),
+		kgo.FetchMaxBytes(100_000_000),
+		kgo.FetchMaxWait(5*time.Second),
+		kgo.FetchMaxPartitionBytes(50_000_000),
+		// Frequently refresh the cluster metadata. This allows us to quickly react in case some brokers are unhealthy,
+		// because the client updates the cluster info only at this frequency (and not after errors too).
+		//
+		// The side effect of frequently refreshing the cluster metadata is that if the backend returns each time a
+		// different authoritative owner for a partition, then each time cluster metadata is updated the Kafka client
+		// will client a new connection for each partition, leading to an high connections churn rate.
+		kgo.MetadataMaxAge(10*time.Second),
+		kgo.WithHooks(r.metrics.kprom),
+		kgo.WithLogger(newKafkaLogger(r.logger)),
+	)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "creating kafka client")
+	}
+
+	return client, nil
+}
+
+func (r *PartitionReader) fetchLastCommittedOffsetWithRetries(ctx context.Context) (offset kgo.Offset, err error) {
+	var (
+		retry = backoff.New(ctx, backoff.Config{
+			MinBackoff: 100 * time.Millisecond,
+			MaxBackoff: 2 * time.Second,
+			MaxRetries: 10,
+		})
+	)
+
+	for retry.Ongoing() {
+		offset, err = r.fetchLastCommittedOffset(ctx)
+		if err == nil {
+			return offset, nil
+		}
+
+		level.Warn(r.logger).Log("msg", "failed to fetch last committed offset", "partition", r.partitionID, "err", err)
+		retry.Wait()
+	}
+
+	// Handle the case the context was canceled before the first attempt.
+	if err == nil {
+		err = retry.Err()
+	}
+
+	return offset, err
+}
+
+func (r *PartitionReader) fetchLastCommittedOffset(ctx context.Context) (kgo.Offset, error) {
+	cl, err := kgo.NewClient(kgo.SeedBrokers(r.kafkaAddress))
+	if err != nil {
+		return kgo.NewOffset(), errors.Wrap(err, "unable to create admin client")
+	}
+	adm := kadm.NewClient(cl)
+	defer adm.Close()
+
+	offsets, err := adm.FetchOffsets(ctx, consumerGroup)
+	if errors.Is(err, kerr.UnknownTopicOrPartition) {
+		// In case we are booting up for the first time ever against this topic.
+		return kgo.NewOffset().AtEnd(), nil
+	}
+	if err != nil {
+		return kgo.NewOffset(), errors.Wrap(err, "unable to fetch group offsets")
+	}
+	offset, _ := offsets.Lookup(r.kafkaTopic, r.partitionID)
+	return kgo.NewOffset().At(offset.At), nil
+}
+
+func (r *PartitionReader) commitLoop(ctx context.Context) {
```

Comment:
[non blocking] Have you considered moving the commitLoop to a dedicated struct with its own service lifecycle, while sharing the same Kafka client instance? It's very isolated from `PartitionReader` and could be cleaner to start/stop as dependency of `PartitionReader`. It would also remove the need of `r.commitLoopWg`.

Reply:
i like how we remove some clutter from ParitionReader with this struct. I moved the pusherConsumer to instantiate its own metrics though, so the struct is no longer shared